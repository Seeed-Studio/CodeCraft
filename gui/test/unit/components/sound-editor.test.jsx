import React from 'react';
import {mountWithIntl, componentWithIntl} from '../../helpers/intl-helpers.jsx';
import SoundEditor from '../../../src/components/sound-editor/sound-editor';

describe('Sound Editor Component', () => {
    let props;
    beforeEach(() => {
        props = {
            canUndo: true,
            canRedo: false,
            chunkLevels: [1, 2, 3],
            name: 'sound name',
            playhead: 0.5,
            trimStart: 0.2,
            trimEnd: 0.8,
            onActivateTrim: jest.fn(),
            onChangeName: jest.fn(),
            onPlay: jest.fn(),
            onRedo: jest.fn(),
            onReverse: jest.fn(),
            onSofter: jest.fn(),
            onLouder: jest.fn(),
            onRobot: jest.fn(),
            onEcho: jest.fn(),
            onFaster: jest.fn(),
            onSlower: jest.fn(),
            onSetTrimEnd: jest.fn(),
            onSetTrimStart: jest.fn(),
            onStop: jest.fn(),
            onUndo: jest.fn()
        };
    });

    test('matches snapshot', () => {
        const component = componentWithIntl(<SoundEditor {...props} />);
        expect(component.toJSON()).toMatchSnapshot();
    });

    test('trim button appears when trims are null', () => {
        const wrapper = mountWithIntl(
            <SoundEditor
                {...props}
                trimEnd={null}
                trimStart={null}
            />
        );
        wrapper.find('[children="Trim"]').simulate('click');
        expect(props.onActivateTrim).toHaveBeenCalled();
    });

    test('save button appears when trims are not null', () => {
        const wrapper = mountWithIntl(
            <SoundEditor
                {...props}
                trimEnd={0.75}
                trimStart={0.25}
            />
        );
        wrapper.find('[children="Save"]').simulate('click');
        expect(props.onActivateTrim).toHaveBeenCalled();
    });

    test('play button appears when playhead is null', () => {
        const wrapper = mountWithIntl(
            <SoundEditor
                {...props}
                playhead={null}
            />
        );
        wrapper.find('button[title="Play"]').simulate('click');
        expect(props.onPlay).toHaveBeenCalled();
    });

    test('stop button appears when playhead is not null', () => {
        const wrapper = mountWithIntl(
            <SoundEditor
                {...props}
                playhead={0.5}
            />
        );
        wrapper.find('button[title="Stop"]').simulate('click');
        expect(props.onStop).toHaveBeenCalled();
    });

    test('submitting name calls the callback', () => {
        const wrapper = mountWithIntl(
            <SoundEditor {...props} />
        );
        wrapper.find('input')
            .simulate('change', {target: {value: 'hello'}})
            .simulate('blur');
        expect(props.onChangeName).toHaveBeenCalled();
    });

    test('effect buttons call the correct callbacks', () => {
        const wrapper = mountWithIntl(
            <SoundEditor {...props} />
        );

        wrapper.find('[children="Reverse"]').simulate('click');
        expect(props.onReverse).toHaveBeenCalled();

        wrapper.find('[children="Echo"]').simulate('click');
        expect(props.onEcho).toHaveBeenCalled();

        wrapper.find('[children="Robot"]').simulate('click');
        expect(props.onRobot).toHaveBeenCalled();

        wrapper.find('[children="Faster"]').simulate('click');
        expect(props.onFaster).toHaveBeenCalled();

        wrapper.find('[children="Slower"]').simulate('click');
        expect(props.onSlower).toHaveBeenCalled();

        wrapper.find('[children="Louder"]').simulate('click');
        expect(props.onLouder).toHaveBeenCalled();

        wrapper.find('[children="Softer"]').simulate('click');
        expect(props.onSofter).toHaveBeenCalled();
    });

    test('undo and redo buttons can be disabled by canUndo/canRedo', () => {
        let wrapper = mountWithIntl(
            <SoundEditor
                {...props}
                canUndo
                canRedo={false}
            />
        );
        expect(wrapper.find('button[title="Undo"]').prop('disabled')).toBe(false);
        expect(wrapper.find('button[title="Redo"]').prop('disabled')).toBe(true);

        wrapper = mountWithIntl(
            <SoundEditor
                {...props}
                canRedo
                canUndo={false}
            />
        );
        expect(wrapper.find('button[title="Undo"]').prop('disabled')).toBe(true);
        expect(wrapper.find('button[title="Redo"]').prop('disabled')).toBe(false);
    });

    test.skip('undo/redo buttons call the correct callback', () => {
        const wrapper = mountWithIntl(
            <SoundEditor
                {...props}
                canRedo
                canUndo
            />
        );
        wrapper.find('button[title="Undo"]').simulate('click');
        expect(props.onUndo).toHaveBeenCalled();

        wrapper.find('button[title="Redo"]').simulate('click');
        expect(props.onRedo).toHaveBeenCalled();
    });
});
