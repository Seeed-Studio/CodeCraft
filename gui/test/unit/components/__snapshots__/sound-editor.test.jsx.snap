// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sound Editor Component matches snapshot 1`] = `
<div
  className={undefined}
>
  <div
    className={undefined}
  >
    <div
      className={undefined}
    >
      <label
        className={undefined}
      >
        <span
          className={undefined}
        >
          Sound
        </span>
        <input
          className=""
          onBlur={[Function]}
          onChange={[Function]}
          onKeyPress={[Function]}
          onSubmit={[Function]}
          tabIndex="1"
          type="text"
          value="sound name"
        />
      </label>
      <div
        className={undefined}
      >
        <button
          className={undefined}
          disabled={false}
          onClick={[Function]}
          title="Undo"
        >
          <img
            className={undefined}
            draggable={false}
            src="test-file-stub"
          />
        </button>
        <button
          className={undefined}
          disabled={true}
          onClick={[Function]}
          title="Redo"
        >
          <img
            className={undefined}
            draggable={false}
            src="test-file-stub"
          />
        </button>
      </div>
    </div>
    <div
      className="undefined"
      onClick={[Function]}
      role="button"
    >
      <img
        className={undefined}
        draggable={false}
        src="test-file-stub"
      />
      <div
        className={undefined}
      >
        <span>
          Save
        </span>
      </div>
    </div>
  </div>
  <div
    className={undefined}
  >
    <div
      className={undefined}
    >
      <svg
        className={undefined}
        viewBox="-1 0 600 160"
      >
        <line
          className={undefined}
          x1={-1}
          x2={600}
          y1={80}
          y2={80}
        />
        <g
          transform="scale(1, -1) translate(0, -80)"
        >
          <path
            className={undefined}
            d="M0 0Q0 0 60 40 Q120 80 180 120 Q240 160 300 200 Q360 240 420 120 Q480 0 480 0 Q480 0 420 -120 Q360 -240 300 -200 Q240 -160 180 -120 Q120 -80 60 -40 Q0 0 0 0Z"
            strokeLinejoin="round"
            strokeWidth={1}
          />
        </g>
      </svg>
      <div
        className={undefined}
      >
        <div
          className=""
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "alignContent": undefined,
              "alignItems": undefined,
              "alignSelf": undefined,
              "flexBasis": undefined,
              "flexDirection": undefined,
              "flexGrow": undefined,
              "flexShrink": undefined,
              "flexWrap": undefined,
              "height": undefined,
              "justifyContent": undefined,
              "width": "20%",
            }
          }
        >
          <div
            className=""
            style={
              Object {
                "alignContent": undefined,
                "alignItems": undefined,
                "alignSelf": undefined,
                "flexBasis": undefined,
                "flexDirection": undefined,
                "flexGrow": undefined,
                "flexShrink": undefined,
                "flexWrap": undefined,
                "height": undefined,
                "justifyContent": undefined,
                "width": undefined,
              }
            }
          />
          <div
            className=""
            style={
              Object {
                "alignContent": undefined,
                "alignItems": undefined,
                "alignSelf": undefined,
                "flexBasis": undefined,
                "flexDirection": undefined,
                "flexGrow": undefined,
                "flexShrink": undefined,
                "flexWrap": undefined,
                "height": undefined,
                "justifyContent": undefined,
                "width": undefined,
              }
            }
          >
            <div
              className=""
              style={
                Object {
                  "alignContent": undefined,
                  "alignItems": undefined,
                  "alignSelf": undefined,
                  "flexBasis": undefined,
                  "flexDirection": undefined,
                  "flexGrow": undefined,
                  "flexShrink": undefined,
                  "flexWrap": undefined,
                  "height": undefined,
                  "justifyContent": undefined,
                  "width": undefined,
                }
              }
            >
              <img
                src="test-file-stub"
              />
            </div>
            <div
              className=""
              style={
                Object {
                  "alignContent": undefined,
                  "alignItems": undefined,
                  "alignSelf": undefined,
                  "flexBasis": undefined,
                  "flexDirection": undefined,
                  "flexGrow": undefined,
                  "flexShrink": undefined,
                  "flexWrap": undefined,
                  "height": undefined,
                  "justifyContent": undefined,
                  "width": undefined,
                }
              }
            >
              <img
                src="test-file-stub"
              />
            </div>
          </div>
        </div>
        <div
          className=""
          style={
            Object {
              "alignContent": undefined,
              "alignItems": undefined,
              "alignSelf": undefined,
              "flexBasis": undefined,
              "flexDirection": undefined,
              "flexGrow": undefined,
              "flexShrink": undefined,
              "flexWrap": undefined,
              "height": undefined,
              "justifyContent": undefined,
              "left": "50%",
              "width": undefined,
            }
          }
        />
        <div
          className=""
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "alignContent": undefined,
              "alignItems": undefined,
              "alignSelf": undefined,
              "flexBasis": undefined,
              "flexDirection": undefined,
              "flexGrow": undefined,
              "flexShrink": undefined,
              "flexWrap": undefined,
              "height": undefined,
              "justifyContent": undefined,
              "left": "80%",
              "width": "20%",
            }
          }
        >
          <div
            className=""
            style={
              Object {
                "alignContent": undefined,
                "alignItems": undefined,
                "alignSelf": undefined,
                "flexBasis": undefined,
                "flexDirection": undefined,
                "flexGrow": undefined,
                "flexShrink": undefined,
                "flexWrap": undefined,
                "height": undefined,
                "justifyContent": undefined,
                "width": undefined,
              }
            }
          />
          <div
            className=""
            style={
              Object {
                "alignContent": undefined,
                "alignItems": undefined,
                "alignSelf": undefined,
                "flexBasis": undefined,
                "flexDirection": undefined,
                "flexGrow": undefined,
                "flexShrink": undefined,
                "flexWrap": undefined,
                "height": undefined,
                "justifyContent": undefined,
                "width": undefined,
              }
            }
          >
            <div
              className=""
              style={
                Object {
                  "alignContent": undefined,
                  "alignItems": undefined,
                  "alignSelf": undefined,
                  "flexBasis": undefined,
                  "flexDirection": undefined,
                  "flexGrow": undefined,
                  "flexShrink": undefined,
                  "flexWrap": undefined,
                  "height": undefined,
                  "justifyContent": undefined,
                  "width": undefined,
                }
              }
            >
              <img
                src="test-file-stub"
              />
            </div>
            <div
              className=""
              style={
                Object {
                  "alignContent": undefined,
                  "alignItems": undefined,
                  "alignSelf": undefined,
                  "flexBasis": undefined,
                  "flexDirection": undefined,
                  "flexGrow": undefined,
                  "flexShrink": undefined,
                  "flexWrap": undefined,
                  "height": undefined,
                  "justifyContent": undefined,
                  "width": undefined,
                }
              }
            >
              <img
                src="test-file-stub"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className=""
  >
    <div
      className={undefined}
    >
      <button
        className=""
        onClick={[Function]}
        title="Stop"
      >
        <img
          draggable={false}
          src="test-file-stub"
        />
      </button>
    </div>
    <div
      className=""
      onClick={[Function]}
      role="button"
    >
      <img
        className={undefined}
        draggable={false}
        src="test-file-stub"
      />
      <div
        className={undefined}
      >
        <span>
          Faster
        </span>
      </div>
    </div>
    <div
      className=""
      onClick={[Function]}
      role="button"
    >
      <img
        className={undefined}
        draggable={false}
        src="test-file-stub"
      />
      <div
        className={undefined}
      >
        <span>
          Slower
        </span>
      </div>
    </div>
    <div
      className=""
      onClick={[Function]}
      role="button"
    >
      <img
        className={undefined}
        draggable={false}
        src="test-file-stub"
      />
      <div
        className={undefined}
      >
        <span>
          Echo
        </span>
      </div>
    </div>
    <div
      className=""
      onClick={[Function]}
      role="button"
    >
      <img
        className={undefined}
        draggable={false}
        src="test-file-stub"
      />
      <div
        className={undefined}
      >
        <span>
          Robot
        </span>
      </div>
    </div>
    <div
      className=""
      onClick={[Function]}
      role="button"
    >
      <img
        className={undefined}
        draggable={false}
        src="test-file-stub"
      />
      <div
        className={undefined}
      >
        <span>
          Louder
        </span>
      </div>
    </div>
    <div
      className=""
      onClick={[Function]}
      role="button"
    >
      <img
        className={undefined}
        draggable={false}
        src="test-file-stub"
      />
      <div
        className={undefined}
      >
        <span>
          Softer
        </span>
      </div>
    </div>
    <div
      className=""
      onClick={[Function]}
      role="button"
    >
      <img
        className={undefined}
        draggable={false}
        src="test-file-stub"
      />
      <div
        className={undefined}
      >
        <span>
          Reverse
        </span>
      </div>
    </div>
  </div>
</div>
`;
